app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt

```diff
@@ -1127,6 +1139,91 @@ class HomeFragment : Fragment() {
         binding.sessionControlRecyclerView.adapter?.notifyDataSetChanged()
     }
 
+    internal fun onPocketStoriesShown(storiesShown: List<PocketRecommendedStory>) {
+        lifecycleScope.launch(IO) {
+            // Cache on disk the updated values for the next time the user opens the homescreen.
+            requireComponents.core.pocketStoriesService.updateStoriesTimesShown(
+                storiesShown.map {
+                    it.copy(timesShown = it.timesShown.inc())
+                }
+            )
+
+            // Update in memory for the next time stories shown are changed (eg: following a category click).
+            var currentCategories = homeFragmentStore.state.pocketStoriesCategories
+            storiesShown.forEach { shownStory ->
+                currentCategories = currentCategories.map { category ->
+                    when (category.name == shownStory.category) {
+                        true -> {
+                            category.copy(
+                                stories = category.stories.map { story ->
+                                    when (story.title == shownStory.title) {
+                                        true -> story.copy(timesShown = story.timesShown.inc())
+                                        false -> story
+                                    }
+                                }
+                            )
+                        }
+                        false -> category
+                    }
+                }
+            }
+
+            homeFragmentStore.dispatch(HomeFragmentAction.PocketStoriesCategoriesChange(currentCategories))
+        }
```

Comment:
I feel like there's too much happening here in the fragment:

* I want to bring up the idea of introducing a Middleware again: Ideally, I think, we should only dispatch an action here, saying "Those stories where shown to the user". The reducer updates the state accordingly. A Middleware may see this action and decide to persist this too.
* Even without the Middleware I think we should update the state in the reducer and not outside and just "injecting" a new state via an action.

Reply:
Updated.